### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by shujaatTracebloc**
## Type of change
- [ ] Feature
- [ ] Bug fix
- [ ] Tech-debt / refactor
- [ ] Docs
- [ ] Security / hardening
- [ ] Breaking change

## Checklist
- [ ] Tests added / updated and passing locally
- [ ] Docs updated if behavior or config changed
- [ ] No secrets / credentials in the diff
- [ ] For security-sensitive paths: appropriate reviewer requested

---

> [!NOTE]
> **Medium Risk**
> Large surface area of new Hub/timm templates (downloads, licenses, gated models) and removal of roberta_base despite README still referencing it; federated/BN and detection head-swap patterns need validation in real training runs.
> 
> **Overview**
> This PR **massively expands the tracebloc model zoo** with modern templates across vision, language, tabular, forecasting, and survival tasks, while tightening **federated-training** and **SDK contract** behavior.
> 
> **Repo / CI:** CI now runs on **`develop`** as well as `master`; `.claude/` is gitignored. **`CLAUDE.md`** adds recommended **`license` metadata**, plus guidance on **BatchNorm**, **EMA buffers**, and **LoRA-only** fine-tuning for large backbones.
> 
> **New templates (high level):** Modern **image classifiers** (e.g. DINOv3, SigLIP2, ConvNeXt V2, timm/HF variants, often frozen backbone + small head). **Object detection** DETR family (DETR, Deformable DETR, RT-DETR, D-FINE, Grounding DINO, OWLv2) plus **FCOS / RetinaNet / Mask R-CNN** with head-swap patterns for custom class counts. **Segmentation** (Mask2Former wrapper for per-pixel CE, SegFormer, SAM2/DINOv3 heads, DPT, etc.). **Keypoint** models (ViTPose, RTMPose, Keypoint R-CNN, Sapiens+LoRA) and **AlphaPose** fixes for configurable joints and a proper `Linear` head for averaging. **MLM** adds many scratch/HF encoders and standardizes **vocab 30522**, position clamping, and `enable_nested_tensor=False`. **Text classification** adds ELECTRA, ModernBERT, Gemma-2, Phi-3, etc.; **`roberta_base.py` is removed**. **Tabular / survival / forecasting** add FT-Transformer, TabM, EBM, HistGradientBoosting, PatchTST, TimesFM+LoRA, TFT, DeepHit/SurvTRACE, and more.
> 
> **Contract tests:** `OPTIONAL_THIRD_PARTY` grows (`peft`, `timm`, `interpret`, etc.) so imports can skip when deps are missing in a CI job.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d2c9d730784d282bff5a77983436fbaf6b3e422d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: model_zoo/keypoint_detection/pytorch/alphapose.py [hunk 1]
```diff
@@ -282,9 +285,10 @@ def __init__(self, norm_layer=nn.BatchNorm2d, cfg=config):
             self.deconv_dim[2], self._preset_cfg['NUM_JOINTS'], kernel_size=1, stride=1, padding=0)
         self.global_avg_pool = nn.AdaptiveAvgPool2d((1, 1))
 
-        # Add a fully connected layer to predict keypoints' x, y coordinates and visibility
-        num_features = self.deconv_dim[2]
-        self.fc = nn.Linear(16, self._preset_cfg['NUM_JOINTS'] * 3)
+        # LazyLinear materializes its weight on the first forward pass once the
+        # flattened feature size is known, but is registered as a proper submodule
+        # immediately — required for state_dict() / federated averaging.
+        self.fc = nn.LazyLinear(self._preset_cfg['NUM_JOINTS'] * 3)
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/keypoint_detection/pytorch/alphapose.py]: Addressed in #80 (already merged into develop): replaced `nn.LazyLinear` with `nn.Linear(num_joints, num_joints * 3)` since the flattened input width is fixed at construction.

File: model_zoo/time_series_forecasting/pytorch/timesfm.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        b, L, n = past_values.shape
+        # (B, L, N) → (B*N, L)
+        flat = past_values.permute(0, 2, 1).reshape(b * n, L)
+        if hasattr(self.base, "forecast"):
+            pred = self.base.forecast(flat, horizon=self.h)
+        else:
+            out = self.base(flat)
+            pred = out.last_hidden_state if hasattr(out, "last_hidden_state") else out
+            pred = pred[..., -self.h:] if pred.ndim == 2 else pred.mean(-1)[..., -self.h:]
+        pred = torch.as_tensor(pred).reshape(b, n, self.h).permute(0, 2, 1)
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/time_series_forecasting/pytorch/timesfm.py]: Addressed in #80 (already merged into develop): dropped the broken generic-forward fallback; the wrapper now unwraps PEFT to reach TimesFM's native `forecast` and raises loudly if unavailable. Autograd issue addressed separately in #82.

File: model_zoo/time_series_forecasting/pytorch/patch_tst.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        prediction_length=forecast_horizon,
+        patch_length=16,
+        patch_stride=8,
+        d_model=128,
+        num_attention_heads=4,
+        num_hidden_layers=3,
+        ffn_dim=256,
+        dropout=0.2,
+    )
+    return PatchTSTForPrediction(config)
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/time_series_forecasting/pytorch/patch_tst.py:30]: Addressed in #82: `MyModel()` in both patch_tst.py and patch_tsmixer.py now accepts `num_feature_points`, `sequence_length`, and `forecast_horizon` kwargs.

File: model_zoo/time_series_forecasting/pytorch/timesfm.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            if hasattr(inner, attr):
+                inner = getattr(inner, attr)
+        if not hasattr(inner, "forecast"):
+            raise AttributeError(
+                "Wrapped TimesFM model does not expose a `forecast` method; "
+                "the zoo's wrapper relies on TimesFM's native forecast API."
+            )
+        pred = inner.forecast(flat, horizon=self.h)
+        pred = torch.as_tensor(pred).reshape(b, n, self.h).permute(0, 2, 1)
+        return pred
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/time_series_forecasting/pytorch/timesfm.py:46]: Addressed in #82: when `self.training` is True, forward now routes through the base model's differentiable `__call__` (PEFT → base.forward) so LoRA adapters receive gradients. `forecast` (no_grad) is used only in eval mode.

File: model_zoo/masked_language_modeling/pytorch/wide_mini_mlm.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self):
+        config = BertConfig(
+            hidden_size=512,
+            num_hidden_layers=4,
+            num_attention_heads=8,
+            intermediate_size=2048,
+            vocab_size=vocab_size,
+            max_position_embeddings=512,
+        )
+        super().__init__(config)
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/masked_language_modeling/pytorch/wide_mini_mlm.py:29]: Addressed in #82: `WideMiniMLM.__init__` now accepts a `vocab_size` kwarg defaulting to the module-level value.

File: tests/test_model_contract.py [hunk 6]
```diff
@@ -26,7 +26,17 @@
...
+    "xgboost",
+    "lightgbm",
+    "catboost",
+    "interpret",
+    "pytorch_forecasting",
+    "pycox",
+    "peft",
+    "tabpfn",
+    "ultralytics",
+}
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply tests/test_model_contract.py:40]: Addressed in #82: added `timm` to `OPTIONAL_THIRD_PARTY` so convnext_v2, eva_02, fastvit, mambavision skip cleanly when timm isn't installed.

File: model_zoo/keypoint_detection/pytorch/vitpose_plus.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        b, k, h, w = heatmaps.shape
+        flat = heatmaps.view(b, k, -1)
+        probs = torch.softmax(flat, dim=-1)
+        ys = torch.linspace(0, 1, h, device=heatmaps.device).view(1, 1, h, 1)
+        xs = torch.linspace(0, 1, w, device=heatmaps.device).view(1, 1, 1, w)
+        probs2d = probs.view(b, k, h, w)
+        x_coord = (probs2d * xs).sum(dim=(2, 3))
+        y_coord = (probs2d * ys).sum(dim=(2, 3))
+        conf = flat.max(dim=-1).values
+        return torch.stack([x_coord, y_coord, conf], dim=-1)
```
[the remark this answers is not in the document]
Comment (divyasinghds) [reply model_zoo/keypoint_detection/pytorch/vitpose_plus.py:35]: Declining: each file in this repo is uploaded standalone via `user.uploadModel(path)` (see CLAUDE.md), so a shared utility module would not be reachable from a single-file upload. Duplication across `vitpose.py` / `vitpose_plus.py` is intentional under that contract.

**Review by LukasWodka — APPROVED** [review]
Approving — Tuesday's develop → master promotion. ruff + test-pytorch + test-sklearn + advance all green. Merging next.
